engine/src/main/java/org/pentaho/di/trans/steps/olapinput/OlapHelper.java

```diff
@@ -0,0 +1,191 @@
+/*! ******************************************************************************
+ *
+ * Pentaho Data Integration
+ *
+ * Copyright (C) 2002-2017 by Hitachi Vantara : http://www.pentaho.com
+ *
+ *******************************************************************************
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.trans.steps.olapinput;
+
+import java.sql.DriverManager;
+import java.sql.SQLException;
+
+import org.olap4j.CellSet;
+import org.olap4j.OlapConnection;
+import org.olap4j.OlapException;
+import org.olap4j.OlapStatement;
+import org.olap4j.OlapWrapper;
+import org.pentaho.di.core.exception.KettleDatabaseException;
+import org.pentaho.di.core.util.Utils;
+import org.pentaho.di.trans.steps.olapinput.olap4jhelper.AbstractBaseCell;
+import org.pentaho.di.trans.steps.olapinput.olap4jhelper.CellDataSet;
+import org.pentaho.di.trans.steps.olapinput.olap4jhelper.OlapUtil;
+
+/**
+ * <code>Olap4j Helper class</code> ...
+ *
+ * @author Paul Stoellberger
+ * @since Mar 12, 2010
+ */
+public class OlapHelper {
+
+  private String olap4jDriver;
+  private String olap4jUrl;
+  private String catalogName;
+  private String mdx;
+  private String username;
+  private String password;
+
+  private CellSet result;
+
+  private String[] headerValues = null;
+  private String[][] cellValues = null;
+  private OlapConnection olapConnection;
+
+  public OlapHelper( String olap4jDriver, String olap4jUrl, String username, String password, String catalogName,
+    String mdx ) {
+    this.olap4jDriver = olap4jDriver;
+    this.olap4jUrl = "jdbc:xmla:Server=" + olap4jUrl;
+    this.catalogName = catalogName;
+    this.mdx = mdx;
+    this.username = username;
+    this.password = password;
+  }
+
+  public void openQuery() throws Exception {
+
+    Class.forName( olap4jDriver );
+    OlapConnection connection = null;
+
+    if ( Utils.isEmpty( username ) && Utils.isEmpty( password ) ) {
+      connection = (OlapConnection) DriverManager.getConnection( olap4jUrl );
+    } else {
+      connection = (OlapConnection) DriverManager.getConnection( olap4jUrl, username, password );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *getConnection*. Currently, there are execution paths that do not contain closure statements, for example, when OlapWrapper.unwrap throws an exception. Ensure the object returned by *getConnection()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).